File: DeckOfCards-Java/src/test/java/com/aprearo/deckofcards/DeckTests.java
import static org.junit.Assert.*;

import org.junit.Before;
import org.junit.Test;

import com.aprearo.deckofcards.*;

import java.util.ArrayList;

public class DeckTests {
    private Deck deck;

    @Before
    public void setUp() throws Exception {
        deck = new Deck();
    }

    @Test
    public void testShuffle() {
        ArrayList<Card> cards1 = new ArrayList<>(deck.getRemainingCards());
        deck.shuffle();
        ArrayList<Card> cards2 = new ArrayList<>(deck.getRemainingCards());
        assertNotEquals(cards1, cards2);
    }

    @Test
    public void testDealOne() {
        int len0 = deck.size();
        Card card1 = deck.dealOne();
        int len1 = deck.size();
        Card card2 = deck.dealOne();
        int len2 = deck.size();
        assertNotEquals(card1, card2);
        assertEquals(len0 - len1, 1);
        assertEquals(len1 - len2, 1);
    }

    @Test(expected=RuntimeException.class)
    public void testNoMoreCards() {
        int size = deck.size();
        for (int i=0; i < size; ++i) {
            deck.dealOne();
        }
        Card card = deck.dealOne();
    }

    @Test
    public void testSameCard() {
        Card card1 = deck.getCard(1);
        Card card2 = deck.getCard(1);
        assertEquals(card1, card2);
        Card card3 = new Card(card1.getSuit(), card1.getRank());
        assertEquals(card1, card3);
    }

    @Test
    public void testDifferentCard() {
        Card card1 = deck.getCard(1);
        Card card2 = deck.getCard(2);
        assertNotEquals(card1, card2);
    }
}

File: DeckOfCards-Kotlin/src/test/kotlin/com/aprearo/deckofcards/DeckTests.kt
package com.aprearo.deckofcards

import org.junit.Before
import org.junit.Test
import kotlin.test.*

class DeckTests {
    private var deck: Deck? = null

    @Before
    fun setUp() {
        deck = Deck()
    }

    @Test
    fun testShuffle() {
        val cards1: List<Card> = deck!!.cards
        deck!!.shuffle()
        val cards2: List<Card> = deck!!.cards
        assertNotEquals(cards1, cards2)
    }

    @Test
    fun testDealOne() {
        val len0 = deck!!.size
        val card1 = deck!!.dealOne()
        val len1 = deck!!.size
        val card2 = deck!!.dealOne()
        val len2 = deck!!.size
        assertNotEquals(card1, card2)
        assertEquals(len0 - len1, 1)
        assertEquals(len1 - len2, 1)
    }

    @Test(expected = RuntimeException::class)
    fun testNoMoreCards() {
        val size = deck!!.size
        for (i in 0 until size) {
            deck!!.dealOne()
        }
        val card = deck!!.dealOne()
    }

    @Test
    fun testSameCard() {
        val card1 = deck!!.getCard(1)
        val card2 = deck!!.getCard(1)
        assertEquals(card1, card2)
        val card3 = Card(card1.suit, card1.rank)
        assertEquals(card1, card3)
    }

    @Test
    fun testDifferentCard() {
        val card1 = deck!!.getCard(1)
        val card2 = deck!!.getCard(2)
        assertNotEquals(card1, card2)
    }
}

